src/tools/ceph_dedup_tool.cc

```diff
@@ -502,6 +514,564 @@ void ChunkScrub::chunk_scrub_common()
+
+    private:
+      size_t dedup_threshold = -1;
+      std::unordered_map<std::string, fp_store_entry_t> fp_map; 
+      std::shared_mutex fingerprint_lock;
+  };
+
+  std::set<std::string> oid_for_evict;
+  static FpStore fp_store;
+  static std::unordered_set<std::string> flushed_objects;
+  static std::shared_mutex flushed_lock;
+  size_t chunk_size;
+  pg_pool_t::fingerprint_t fp_type;
+};
+
+SampleDedup::FpStore SampleDedup::fp_store;
+SampleDedup::crawl_mode_t default_crawl_mode = SampleDedup::crawl_mode_t::DEEP;
+std::unordered_set<std::string> SampleDedup::flushed_objects;
+std::shared_mutex SampleDedup::flushed_lock;
+
+void SampleDedup::crawl() {
+  try {
+    if (fp_type == pg_pool_t::TYPE_FINGERPRINT_NONE) {
+      throw std::logic_error("unknown fingerprint algorithm");
+    }
+    prepare_rados();
+    ObjectCursor shard_start;
+    ObjectCursor shard_end;
+    std::tie(shard_start, shard_end) = get_shard_boundary();
+    cout << "new iteration thread: " << n <<std::endl;
+
+    for (ObjectCursor current_object = shard_start;
+        current_object < shard_end;) {
+      std::vector<ObjectItem> objects;
+      // Get the list of object IDs to deduplicate
+      std::tie(objects, current_object) = get_objects(
+          current_object,
+          shard_end,
+          100);
+
+      // Pick few objects to be processed. Crawling mode decides how many
+      // objects to pick (sampling ratio). Lower sampling ratio makes crawler
+      // have lower crawling overhead but find less duplication.
+      std::set<size_t> sampled_indexes = sample_object(objects.size());
+      for (size_t index : sampled_indexes) {
+        ObjectItem target = objects[index];
+        // Only process dirty objects which are expected not processed yet
+        if (is_dirty(target)) {
+          try_object_dedup_and_accumulate_result(target);
+        }
+      }
+    }
+
+    map<std::string,AioCompletion*> set_chunk_completions;
+    // Do set_chunk to make found duplicable chunks can be evicted by tier_evict()
+    for (auto& duplicable_chunk : duplicable_chunks) {
+      auto completion = set_chunk_duplicated(duplicable_chunk);
+      if (completion != nullptr) {
+        set_chunk_completions[duplicable_chunk.oid] = completion;
+      }
+    }
+
+    vector<AioCompletion*> evict_completions;
+    for (auto& oid : oid_for_evict) {
+      auto completion_iter = set_chunk_completions.find(oid);
+      // Related set_chunk should be completed before tier_evict because
+      // tier_evict() only evict data processed by set_chunk() or tier_flush()
+      if (completion_iter != set_chunk_completions.end()) {
+        auto completion = completion_iter->second;
+        completion->wait_for_complete();
+        delete completion;
+      }
+      auto completion = do_async_evict(oid);
+      evict_completions.push_back(completion);
+    }
+    for (auto& completion : evict_completions) {
+      completion->wait_for_complete();
+      delete completion;
```

Comment:
We really want raw pointers to be wrapped by smart pointers whenever possible.  Modify evict_completions and set_chunk_completions to be std::vector<std::unique_ptr<AioCompletion>> and return std::unique<AioCompletion> from do_async_evict and set_chunk_duplicated.